### PR TITLE
fix(kafka source): Fix handling of negative acknowledgements

### PR DESCRIPTION
**Pull request by bruceg**
This change improves the handling of negative end-to-end acknowledgements in the Kafka source by stopping updates on a topic with a nack and then stopping further event deliveries for that topic.

@binarylogic @jszwedko Could you confirm the above behavior is the desired end goal? It is what I recall of our last discussion on the issue but differs for the note made in the issue #12109. I made the changes in separate steps so I could revert the behavior changes if needed.

### CODE REVIEW

**Review by blt — APPROVED** [review]
Looks straightforward to me. Glad to see some magic numbers moved into constants.

**Review by jszwedko — COMMENTED** [review]
👍 the behavior seems right to me. It seems to match [the note you referenced](https://github.com/vectordotdev/vector/issues/12109#issuecomment-1093200417) though so now I'm worried I don't understand it if the behavior here does actually differ 😅  Does it differ from that note? If so, how so?

My only other thought would be including a log message at the ERROR level when this happens as users are very likely to want to intervene.

**Comment by bruceg** [comment]
> 👍 the behavior seems right to me. It seems to match [the note you referenced](https://github.com/vectordotdev/vector/issues/12109#issuecomment-1093200417) though so now I'm worried I don't understand it if the behavior here does actually differ 😅 Does it differ from that note? If so, how so?

The comment on #12109 says:
> we have decided to make these sources stop updating the checkpoint on whichever stream receives a negative acknowledgement

Which implies that they will continue processing but just not update the checkpoint. That was different from what I recalled discussing, which was to also stop handling new events. If that is your recollection as well, then I will clarify the comment.

> My only other thought would be including a log message at the ERROR level when this happens as users are very likely to want to intervene.

Agreed. I will add an internal event for this.

**Comment by jszwedko** [comment]
> Which implies that they will continue processing but just not update the checkpoint. That was different from what I recalled discussing, which was to also stop handling new events. If that is your recollection as well, then I will clarify the comment.
> 
> >

Aha, I see. I think either behavior would have been acceptable, but what you have here is preferable to reduce duplicate events if Vector is restarted after the operator fixes the issue.

File: src/sources/kafka.rs [hunk 1]
```diff
@@ -255,16 +295,21 @@ async fn parse_message(
...
     keys: Keys<'a>,
+    topics: &Topics,
 ) -> Option<(usize, impl Stream<Item = Event> + 'a)> {
     let payload = msg.payload()?; // skip messages with empty payload
 
     let rmsg = ReceivedMessage::from(msg);
 
+    if topics.failed.contains(&rmsg.topic) {
+        return None;
+    }
```
Comment (tobz) [inline src/sources/kafka.rs:308]: This should be at the very top of the function, it seems like.

If we know we're going to skip any queued messages that we hadn't yet drained when marking a topic as failed, we might as well be as efficient as possible about it.

**Review by tobz — APPROVED** [review]
Looks good to me overall.

There's some questions, suggestions, but it's all non-blocking (albeit recommended).

File: src/sources/kafka.rs [hunk 2]
```diff
@@ -179,17 +179,13 @@ async fn kafka_source(
...
             _ = &mut shutdown => break,
             entry = ack_stream.next() => if let Some((status, entry)) = entry {
-                if status == BatchStatus::Delivered {
-                    if let Err(error) =
-                        consumer.store_offset(&entry.topic, entry.partition, entry.offset)
-                    {
-                        emit!(KafkaOffsetUpdateError { error });
-                    }
-                }
+                handle_ack(&mut topics, status, entry,&consumer);
```
[the remark this answers is not in the document]
Comment (bruceg) [reply src/sources/kafka.rs:188]: It's in a `select!` macro, which confuses `rustfmt` sometimes. Will fix.

File: src/sources/kafka.rs (hunk 1 above)
Comment (bruceg) [reply src/sources/kafka.rs:308]: You are right. However, there is a cost there of calling `msg.topic.to_string()` twice when there are no rejected messages. That's probably smaller, though.

File: src/sources/kafka.rs [hunk 3]
```diff
@@ -211,15 +207,59 @@ async fn kafka_source(
...
+                topic: &entry.topic,
+                partition: entry.partition,
+                offset: entry.offset,
+            });
+            // Try to unsubscribe from the named topic
+            if topics.subscribed.remove(&entry.topic) {
+                let topics: Vec<&str> = topics.subscribed.iter().map(|s| s.as_str()).collect();
+                // Ignore error, as we ignore output from the topic below anyways
+                let _ = consumer.subscribe(&topics);
+            }
```
[the remark this answers is not in the document]
Comment (tobz) [reply src/sources/kafka.rs:246]: Makes sense. 👍🏻

File: src/sources/kafka.rs [hunk 4]
```diff
@@ -211,15 +207,59 @@ async fn kafka_source(
...
+            emit!(KafkaNegativeAcknowledgmentError {
+                topic: &entry.topic,
+                partition: entry.partition,
+                offset: entry.offset,
+            });
+            // Try to unsubscribe from the named topic
+            if topics.subscribed.remove(&entry.topic) {
+                let topics: Vec<&str> = topics.subscribed.iter().map(|s| s.as_str()).collect();
+                // Ignore error, as we ignore output from the topic below anyways
+                let _ = consumer.subscribe(&topics);
```
[the remark this answers is not in the document]
Comment (tobz) [reply src/sources/kafka.rs:245]: Hmm... but empirically, it seems to do the logical operation of keeping other topic subscriptions while removing the failed topic?

As long as it does that correctly, we can _probably_ rely on it, but it's still worth a doc comment.

**Review by StephenWakely — APPROVED** [review]
Looks good to me!

File: src/sources/kafka.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (tobz) [reply src/sources/kafka.rs:245]: Ah yeah, nice catch! This seems reasonable then.

Although, still, could use a more detailed comment IMO.

[the remark this answers is not in the document]
Comment (bruceg) [reply src/sources/kafka.rs:245]: Yes, nice catch. I was trying to actually test the behavior of the subscribe function, but had a bug in my code. I will comment more there.

**Pull request merged**